### PR TITLE
perf(sampling): defer logits block reduction

### PR DESCRIPTION
**Pull request by tomalmog**
## Description

`SamplingFromLogitsKernel` currently performs a block reduction and barrier for
every vocabulary tile. This patch keeps a running maximum per vector lane and
reduces once after the loop. For an FP32 vocabulary of 151,936, that replaces
38 block reductions with one.

The RNG, tie ordering, vector-lane reduction layout and padded-output behavior
stay the same. There are no public API changes.

Adds 62 regression cases for masked tile boundaries, int32/int64 indexed rows,
cross-tile ties and CUDA-graph input updates, plus performance and seeded-output
comparison scripts for the public API.

### Performance

Compared with unmodified `93e9eef092c11bc93a6f432a16404054e85d19c5` on A100-SXM4
80GB and H100 80GB, using Torch 2.8.0 / CUDA 12.8 / TVM-FFI 0.1.11 and source JIT
builds targeting SM80/SM90a. FP32 logits, temperature 1, no filtering, scalar
seed/offset. CUDA-graph event timing excludes compilation, allocation and Python
dispatch. Each arm runs sequentially on the same GPU, with 11 timing rounds.

Across B=1/8/32/128/512, V=32000/128256/151936 and hot/128MiB-streamed inputs,
all 60 configurations improve: 8.6-35.3% lower latency on A100 and 21.2-32.0% on H100.
Representative hot-input results:

| GPU | Batch | Vocab | Cache | Before (us) | After (us) | Speedup |
|---|---:|---:|---|---:|---:|---:|
| A100 | 1 | 151936 | hot | 91.989 | 59.972 | 1.534x |
| A100 | 128 | 151936 | hot | 184.047 | 119.569 | 1.539x |
| A100 | 512 | 151936 | hot | 458.445 | 296.482 | 1.546x |
| H100 | 1 | 151936 | hot | 62.293 | 42.775 | 1.456x |
| H100 | 128 | 151936 | hot | 63.399 | 43.780 | 1.448x |
| H100 | 512 | 151936 | hot | 247.218 | 168.175 | 1.470x |

Copy the two added benchmark/check scripts to the baseline checkout as well.
Use an editable/source installation on each revision and separate JIT caches:

```bash
FLASHINFER_WORKSPACE_BASE=/tmp/fi-before \
  python benchmarks/bench_sampling_logits.py --output before.json
FLASHINFER_WORKSPACE_BASE=/tmp/fi-after \
  python benchmarks/bench_sampling_logits.py --output after.json
```

These results apply to the existing logits sampling API. Whole-model throughput
was not measured.

## Related Issues

No tracked issue. Nearby #4861 changes seed resolution; this change leaves that logic untouched.

## Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

Affected tests below passed; the full repository suite was not run.

```bash
python -m pytest -q tests/utils/test_sampling_logits_reduction.py \
  tests/utils/test_sampling.py \
  tests/utils/test_logits_processor.py::TestLogitsPipeVsSamplingOps::test_logits_sample \
  -k 'logits and not top_k and not top_p'
```

110 passed, 1878 deselected on **each** GPU. The 62 new tests also pass on the
unmodified revision. Separately, 576 original-versus-patched seeded cases per GPU
(12,672 token comparisons total) agree exactly, covering random/tied/masked logits,
both deterministic modes, indexed rows and scalar/tensor seed inputs.

For the seeded comparison, run `python benchmarks/check_sampling_logits.py before-seeded.json`
on the baseline and the same script with `after-seeded.json` on the patched source,
then compare the JSON files. Compare within the same GPU, not across architectures.

Applicable pre-commit hooks passed on all four changed files. Pre-commit and its
hook environments ran in a container; no host Git hook was installed and full-tree hooks
were not run (the corresponding checklist boxes remain unchecked).

A broader H100 `-k logits` attempt was interrupted while compiling unrelated
top-k routines, after 86 tests passed. That incomplete attempt is not counted as
a passing broad suite. Earlier native-dispatch memcheck/racecheck runs passed on
both GPUs; no full-model benchmark was performed.

## Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Reviewer Notes

Please review the tie comparator and preserved vector-lane reduction layout.
GPU CI on the other supported architectures remains pending maintainer triggering.
AI assistance was used for implementation, investigation and validation.

### CODE REVIEW

**Comment by tomalmog** [comment]
First PRs here, so I don't have permission to start CI. Could a maintainer run `@flashinfer-bot run` when convenient? This one and #5201 are both small sampling-kernel changes with tests and before/after benchmarks in the descriptions. Happy to rebase either if preferred.
